File: backend/app/routers/institutions.py
"""Membership-scoped institution APIs. No platform-role bypass."""
import csv
import io
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import Response
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.services.auth_service import AuthService
from app.services import institution_service as svc
from app.schemas.institution import (
    InstitutionCreate,
    InviteCreate,
    MemberUpdate,
    BatchCreate,
    BatchMembers,
    CourseConnect,
    AssignmentCreate,
    PlanRequestCreate,
    PlanRequestReview,
)
from app.models.institution import (
    InstitutionInvite,
    InstitutionBatchMember,
    InstitutionPlanRequest,
    Institution,
)
from app.models.user import User

router = APIRouter()
CurrentUser = Depends(AuthService.get_current_active_user)


@router.get("/platform/plan-requests")
def platform_requests(db: Session = Depends(get_db), user=CurrentUser):
    if user.role not in ("admin", "superadmin"):
        raise HTTPException(403, "Platform administrator access required.")
    return [
        {
            "id": r.id,
            "institution": inst.name,
            "institution_id": inst.id,
            "plan": r.plan,
            "status": r.status,
            "note": r.note,
            "contact_email": owner.user_email,
            "created_at": svc.utc(r.created_at),
        }
        for r, inst, owner in db.query(InstitutionPlanRequest, Institution, User)
        .join(Institution, Institution.id == InstitutionPlanRequest.institution_id)
        .join(User, User.id == InstitutionPlanRequest.requested_by)
        .order_by(InstitutionPlanRequest.id.desc())
        .limit(200)
        .all()
    ]


@router.patch("/platform/plan-requests/{request_id}")
def review_request(
    request_id: int,
    data: PlanRequestReview,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    if user.role not in ("admin", "superadmin"):
        raise HTTPException(403, "Platform administrator access required.")
    row = (
        db.query(InstitutionPlanRequest)
        .filter_by(id=request_id)
        .with_for_update()
        .first()
    )
    if not row:
        raise HTTPException(404, "Request not found.")
    if row.status != "pending":
        raise HTTPException(409, "This request has already been reviewed.")
    row.status = data.status
    svc.audit(
        db,
        row.institution_id,
        user,
        "plan.request_reviewed",
        f"{data.status}: {data.note}",
    )
    svc.save(db)
    return {"status": row.status}


@router.get("")
def mine(db: Session = Depends(get_db), user=CurrentUser):
    return svc.list_mine(db, user)


@router.post("", status_code=201)
def create(data: InstitutionCreate, db: Session = Depends(get_db), user=CurrentUser):
    return svc.create(db, user, data)


@router.get("/invitations")
def invitations(db: Session = Depends(get_db), user=CurrentUser):
    return svc.pending_invites(db, user)


@router.post("/invitations/{invite_id}/accept")
def accept(invite_id: int, db: Session = Depends(get_db), user=CurrentUser):
    return svc.accept_invite(db, invite_id, user)


@router.get("/{institution_id}")
def overview(institution_id: int, db: Session = Depends(get_db), user=CurrentUser):
    return svc.overview(db, institution_id, user)


@router.patch("/{institution_id}")
def update(
    institution_id: int,
    data: InstitutionCreate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    inst, member = svc.scope(db, institution_id, user, svc.MANAGERS, lock=True)
    for key, value in data.model_dump().items():
        setattr(inst, key, value)
    svc.audit(db, institution_id, user, "institution.updated", inst.name)
    svc.save(db)
    return svc.serialize(inst, member.role)


@router.post("/{institution_id}/invitations", status_code=201)
def invite(
    institution_id: int,
    data: InviteCreate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    return svc.invite(db, institution_id, user, data)


@router.delete("/{institution_id}/invitations/{invite_id}")
def revoke_invite(
    institution_id: int, invite_id: int, db: Session = Depends(get_db), user=CurrentUser
):
    _, actor = svc.scope(db, institution_id, user, svc.MANAGERS, lock=True)
    row = (
        db.query(InstitutionInvite)
        .filter_by(id=invite_id, institution_id=institution_id)
        .first()
    )
    if not row:
        raise HTTPException(404, "Invitation not found.")
    if row.role == "admin" and actor.role != "owner":
        raise HTTPException(
            403, "Only the owner can revoke an administrator invitation."
        )
    if row.status != "pending":
        raise HTTPException(409, "This invitation is no longer pending.")
    row.status = "revoked"
    svc.audit(db, institution_id, user, "invitation.revoked", row.email)
    svc.save(db)
    return {"status": "revoked"}


@router.patch("/{institution_id}/members/{member_id}")
def member_update(
    institution_id: int,
    member_id: int,
    data: MemberUpdate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.update_member(db, institution_id, member_id, user, data)
    return {"status": "updated"}


@router.post("/{institution_id}/batches", status_code=201)
def batch_create(
    institution_id: int,
    data: BatchCreate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    return svc.create_batch(db, institution_id, user, data)


@router.post("/{institution_id}/batches/{batch_id}/members")
def batch_members(
    institution_id: int,
    batch_id: int,
    data: BatchMembers,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.add_batch_members(db, institution_id, batch_id, user, data.member_ids)
    return {"status": "updated"}


@router.delete("/{institution_id}/batches/{batch_id}/members/{member_id}")
def remove_batch_member(
    institution_id: int,
    batch_id: int,
    member_id: int,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.scope(db, institution_id, user, svc.MANAGERS, lock=True)
    batch = svc.batch_scope(db, institution_id, batch_id)
    row = (
        db.query(InstitutionBatchMember)
        .filter_by(batch_id=batch_id, member_id=member_id)
        .first()
    )
    if not row:
        raise HTTPException(404, "Batch member not found.")
    db.delete(row)
    svc.audit(
        db,
        institution_id,
        user,
        "batch.member_removed",
        f"{batch.name}: member #{member_id}",
    )
    svc.save(db)
    return {"status": "removed"}


@router.get("/{institution_id}/available-courses")
def available(institution_id: int, db: Session = Depends(get_db), user=CurrentUser):
    return svc.available_courses(db, institution_id, user)


@router.post("/{institution_id}/courses", status_code=201)
def connect(
    institution_id: int,
    data: CourseConnect,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.connect_course(db, institution_id, user, data.course_id)
    return {"status": "connected"}


@router.post("/{institution_id}/batches/{batch_id}/assignments")
def assign(
    institution_id: int,
    batch_id: int,
    data: AssignmentCreate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.assign_course(db, institution_id, batch_id, user, data)
    return {"status": "assigned"}


@router.get("/{institution_id}/report.csv")
def export_report(institution_id: int, db: Session = Depends(get_db), user=CurrentUser):
    svc.scope(db, institution_id, user, svc.STAFF)
    rows = svc.report_rows(db, institution_id, user)
    output = io.StringIO(newline="")
    writer = csv.writer(output)
    writer.writerow(
        ["Batch", "Student", "Course", "Progress (%)", "Enrollment", "Due date"]
    )

    def safe(value):
        value = str(value or "")
        return (
            "'" + value
            if value.lstrip().startswith(("=", "+", "-", "@", "\t", "\r"))
            else value
        )

    for row in rows:
        writer.writerow(
            [
                safe(row["batch"]),
                safe(row["student"]),
                safe(row["course"]),
                row["progress"],
                "Enrolled" if row["has_access"] else "Enrollment needed",
                row["due_date"] or "",
            ]
        )
    return Response(
        "\ufeff" + output.getvalue(),
        media_type="text/csv",
        headers={
            "Content-Disposition": f'attachment; filename="institution-{institution_id}-progress.csv"',
            "Cache-Control": "no-store",
        },
    )


@router.post("/{institution_id}/plan-requests", status_code=201)
def request_plan(
    institution_id: int,
    data: PlanRequestCreate,
    db: Session = Depends(get_db),
    user=CurrentUser,
):
    svc.request_plan(db, institution_id, user, data)
    return {
        "status": "pending",
        "detail": "Request saved for review. Your current plan has not changed.",
    }

File: frontend/src/api/institutions.ts
import { api } from "./axios";

export type InstitutionRole = "owner" | "admin" | "teacher" | "student";
export interface Institution {
  id: number;
  name: string;
  slug: string;
  kind: "school" | "college";
  academic_year: string;
  timezone: string;
  description: string;
  plan: string;
  role: InstitutionRole;
}
export interface InstitutionInput {
  name: string;
  kind: "school" | "college";
  academic_year: string;
  timezone: string;
  description: string;
}
export interface InstitutionMember {
  id: number;
  user_id: number;
  name: string;
  email: string;
  role: InstitutionRole;
  department: string;
  status: string;
}
export interface InstitutionBatch {
  id: number;
  name: string;
  department: string;
  academic_year: string;
  student_count: number;
  member_ids: number[];
  assignments: {
    id: number;
    institution_course_id: number;
    due_date: string | null;
  }[];
}
export interface InstitutionCourse {
  id: number;
  course_id: number;
  title: string;
  status: string;
  can_edit: boolean;
}
export interface ProgressRow {
  batch_id: number;
  batch: string;
  student: string;
  user_id: number;
  course_id: number;
  course: string;
  due_date: string | null;
  progress: number;
  has_access: boolean;
}
export interface PlatformPlanRequest {
  id: number;
  institution: string;
  institution_id: number;
  plan: string;
  status: string;
  note: string;
  contact_email: string;
  created_at: string;
}
export interface InstitutionOverview {
  institution: Institution;
  members: InstitutionMember[];
  batches: InstitutionBatch[];
  courses: InstitutionCourse[];
  invites: {
    id: number;
    email: string;
    role: string;
    status: string;
    expires_at: string;
  }[];
  activity: {
    id: number;
    action: string;
    detail: string;
    created_at: string;
  }[];
  usage: {
    members: number;
    reserved_seats: number;
    batches: number;
    courses: number;
    limits: { members: number; batches: number; courses: number };
  } | null;
  plan_requests: {
    id: number;
    plan: string;
    status: string;
    created_at: string;
  }[];
  report: ProgressRow[];
}
const root = "/institutions";
export const institutionApi = {
  platformRequests: async () =>
    (await api.get<PlatformPlanRequest[]>(`${root}/platform/plan-requests`))
      .data,
  reviewRequest: async (id: number, status: string, note: string) =>
    api.patch(`${root}/platform/plan-requests/${id}`, { status, note }),
  list: async () => (await api.get<Institution[]>(root)).data,
  invitations: async () =>
    (
      await api.get<
        {
          id: number;
          institution_name: string;
          role: string;
          expires_at: string;
        }[]
      >(`${root}/invitations`)
    ).data,
  accept: async (id: number) =>
    (await api.post<Institution>(`${root}/invitations/${id}/accept`)).data,
  create: async (data: InstitutionInput) =>
    (await api.post<Institution>(root, data)).data,
  overview: async (id: number) =>
    (await api.get<InstitutionOverview>(`${root}/${id}`)).data,
  update: async (id: number, data: InstitutionInput) =>
    (await api.patch<Institution>(`${root}/${id}`, data)).data,
  invite: async (
    id: number,
    data: { email: string; role: string; department: string },
  ) => api.post(`${root}/${id}/invitations`, data),
  revoke: async (id: number, inviteId: number) =>
    api.delete(`${root}/${id}/invitations/${inviteId}`),
  updateMember: async (
    id: number,
    memberId: number,
    data: { role: string; department: string; status: string },
  ) => api.patch(`${root}/${id}/members/${memberId}`, data),
  batch: async (
    id: number,
    data: { name: string; department: string; academic_year: string },
  ) => api.post(`${root}/${id}/batches`, data),
  addStudents: async (id: number, batchId: number, member_ids: number[]) =>
    api.post(`${root}/${id}/batches/${batchId}/members`, { member_ids }),
  removeStudent: async (id: number, batchId: number, memberId: number) =>
    api.delete(`${root}/${id}/batches/${batchId}/members/${memberId}`),
  available: async (id: number) =>
    (
      await api.get<{ id: number; title: string; status: string }[]>(
        `${root}/${id}/available-courses`,
      )
    ).data,
  connect: async (id: number, course_id: number) =>
    api.post(`${root}/${id}/courses`, { course_id }),
  assign: async (
    id: number,
    batchId: number,
    institution_course_id: number,
    due_date: string | null,
  ) =>
    api.post(`${root}/${id}/batches/${batchId}/assignments`, {
      institution_course_id,
      due_date,
    }),
  requestPlan: async (id: number, plan: string, note: string) =>
    api.post(`${root}/${id}/plan-requests`, { plan, note }),
  export: async (id: number) =>
    (await api.get<Blob>(`${root}/${id}/report.csv`, { responseType: "blob" }))
      .data,
};
